torch/csrc/lazy/core/shape_inference.cpp

```diff
@@ -977,6 +977,89 @@ std::vector<Shape> compute_shape__adaptive_avg_pool2d_backward(
   return {Shape(self.scalar_type(), self.sizes().vec())};
 }
 
+std::vector<Shape> compute_shape__adaptive_avg_pool3d(
+    const at::Tensor& self,
+    at::IntArrayRef output_size) {
+  // Checks based on `aten/src/ATen/native/AdaptiveAveragePooling.cpp`
+  // and on `aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp`
+  TORCH_CHECK(
+      output_size.size() == 3, "adaptive_avg_pool3d: output_size must be 3");
+  TORCH_CHECK(
+      (output_size[0] >= 0 && output_size[1] >= 0 && output_size[2] >= 0),
+      "adaptive_avg_pool3d: elements of output_size must be greater than or equal to 0 ",
+      "but received {",
+      output_size[0],
+      ", ",
+      output_size[1],
+      ", ",
+      output_size[2],
+      "}");
+  int64_t ndim = self.ndimension();
+  for (const auto i : c10::irange(1, ndim)) {
+    TORCH_CHECK(
+        self.size(i) > 0,
+        "adaptive_avg_pool3d(): Expected self to have non-zero size for non-batch dimensions, "
+        "but Tensor has sizes ",
+        self.sizes(),
+        " with dimension ",
+        i,
+        " being "
+        "empty");
+  }
+  TORCH_CHECK(
+      (ndim == 4 || ndim == 5),
```

Comment:
nit: we could consider moving this up as this is one of the most inexpensive checks